packages/test/test-end-to-end-tests/src/test/data-virtualization/groupIdOffline.spec.ts

```diff
@@ -0,0 +1,389 @@
+			testDataObjectType,
+			loadingGroupId,
+		);
+
+		// Attach the data stores
+		const dataObjectA = (await dataStoreA.entryPoint.get()) as TestDataObject;
+		const dataObjectB = (await dataStoreB.entryPoint.get()) as TestDataObject;
+		mainObject._root.set("dataObjectA", dataObjectA.handle);
+		mainObject._root.set("dataObjectB", dataObjectB.handle);
+		await provider.ensureSynchronized();
+
+		const { summarizer } = await createSummarizerFromFactory(
+			provider,
+			container,
+			dataObjectFactory,
+			undefined,
+			undefined,
+			undefined,
+			undefined,
+			configProvider,
+		);
+
+		await provider.ensureSynchronized();
+		const { summaryVersion } = await summarizeNow(summarizer);
+		await provider.ensureSynchronized();
+
+		const container2 = (await provider.loadContainer(
+			runtimeFactory,
+			{ configProvider },
+			{ [LoaderHeader.version]: summaryVersion },
+		)) as IContainerExperimental;
+		await provider.ensureSynchronized();
+		const mainObject2 = (await container2.getEntryPoint()) as TestDataObject;
+		const handleA2 = mainObject2._root.get<IFluidHandle<TestDataObject>>("dataObjectA");
+		const handleB2 = mainObject2._root.get<IFluidHandle<TestDataObject>>("dataObjectB");
+		callCount = 0;
+		const dataObjectA2 = await handleA2?.get();
+		const dataObjectB2 = await handleB2?.get();
+		assert.equal(callCount, 1, "Should have made a network call");
+		assert(dataObjectA2 !== undefined, "dataObjectA2 should not be undefined");
+		assert(dataObjectB2 !== undefined, "dataObjectB2 should not be undefined");
+		dataObjectA2._root.set("A", "A");
+		dataObjectB2._root.set("B", "B");
+		await provider.ensureSynchronized();
+
+		const { summaryRefSeq } = await summarizeNow(summarizer);
+		await provider.ensureSynchronized();
+
+		// Refresh snapshot - this to validate that we built a system that would work with refresh
+		// There are two parts to refresh, making the network snapshot call and trimming the ops
+		// Container layer Refresh
+		// Network call refreshing the base snapshot
+		const serializedStateManager = (container2 as any).serializedStateManager;
+		await serializedStateManager.refreshLatestSnapshot();
+
+		// Update the latestSequenceNumber so that the reference sequence number is beyond the snapshot
+		await provider.ensureSynchronized();
+		container2.disconnect();
+		dataObjectA2._root.set("A2", "A2");
+		dataObjectB2._root.set("B2", "B2");
+
+		// Hack to make sure we don't immediately fail/close the container on pending ops
+		// Another way around this is to simply have a different container send remote messages.
+		// What happens is that the last two synced ops we made are considered "saved", This may be useful for testing an offline edge case
+		// The last two saved ops (setting A and B) have reference sequence numbers that point to a sequence number
+		// before the snapshot
+		(dataObjectA2.containerRuntime as any).pendingStateManager.savedOps = [];
```

Comment:
@anthony-murphy this may be interesting for you to review as this may be relevant to the refresh effort.

cc: @scarlettjlee , I don't know who else in offline mode might find this particularly interesting to review.